docs/proposals/templates.md

```diff
@@ -0,0 +1,550 @@
+          "metadata": {
+              "creationTimestamp": null,
+              "labels": {
+                  "name": "$(DATABASE_SERVICE_NAME)"
+              }
+          },
+          "spec": {
+            "containers": [
+              {
+                "name": "mongodb",
+                "image": "docker.io/centos/mongodb-26-centos7",
+                "ports": [
+                  {
+                    "containerPort": 27017,
+                    "protocol": "TCP"
+                  }
+                ],
+                "env": [
+                  {
+                    "name": "MONGODB_USER",
+                    "value": "${MONGODB_USER}"
+                  },
+                  {
+                    "name": "MONGODB_PASSWORD",
+                    "value": "${MONGODB_PASSWORD}"
+                  },
+                  {
+                    "name": "MONGODB_DATABASE",
+                    "value": "${MONGODB_DATABASE}"
+                  }
+                ]
+              }
+            ]
+          }
+        }
+      }
+    }
+  ],
+  "parameters": [
+    {
+      "name": "DATABASE_SERVICE_NAME",
+      "description": "Database service name",
+      "value": "mongodb",
+      "required": true,
+      "type": "string"
+    },
+    {
+      "name": "MONGODB_USER",
+      "description": "Username for MongoDB user that will be used for accessing the database",
+      "value": "password"
+      "required": true,
+      "type": "string"
+    },
+    {
+      "name": "MONGODB_PASSWORD",
+      "description": "Password for the MongoDB user",
+      "required": true,
+      "type" :"string"
+    },
+    {
+      "name": "MONGODB_DATABASE",
+      "description": "Database name",
+      "value": "sampledb",
+      "required": true,
+      "type" :"string"      
+    },
+    {
+      "name": "REPLICA_COUNT",
+      "description": "Number of mongo replicas to run",
+      "value": "1",
+      "required": true,
+      "type" :"int"      
+    }
+  ]
+}
+```
+
+### API Endpoints
+
+* **/processedTemplates** - when a template is POSTed to this endpoint, all parameters in the template are processed, values 
+generated as necessary, and substituted into appropriate locations in the object definitions.  Validation is performed to ensure
+required parameters have a value (either supplied in the original template or generated by a defined generator).  In addition 
+labels defined in the template are applied to the object definitions.  Finally the customized template (still a `Template` object) 
+is returned to the caller.
+
+Performing parameter substitution on the server side has the benefit of centralizing the processing so that new clients of 
+k8s, such as IDEs, CI systems, Web consoles, etc, do not need to reimplement template processing or embed the k8s binary.
+Instead they can invoke the k8s api directly.
+
+This resource can also support a subresource "/processedTemplates/templatename".  This resource would accept just a Parameters object
```

Comment:
I was thinking more of a subresource of the specific template: `/namespaces/myns/templates/mytemplate/substituted`